test/lib/DuplexCollection.js

```diff
@@ -0,0 +1,138 @@
+const {test} = require("ava");
+const sinon = require("sinon");
+const DuplexCollection = require("../../lib/DuplexCollection");
+const ReaderCollectionPrioritized = require("../../lib/ReaderCollectionPrioritized");
+const Resource = require("../../lib/Resource");
+
+test("DuplexCollection: constructor", (t) => {
+	t.plan(5);
+
+	const duplexCollection = new DuplexCollection({
+		name: "myCollection",
+		reader: {},
+		writer: {}
+	});
+
+	t.deepEqual(duplexCollection._reader, {}, "reader assigned");
+	t.deepEqual(duplexCollection._writer, {}, "writer assigned");
+	t.true(duplexCollection._combo instanceof ReaderCollectionPrioritized, "prioritized reader collection created");
+	t.deepEqual(duplexCollection._combo._name, "myCollection", "name assigned");
+	t.deepEqual(duplexCollection._combo._readers, [{}, {}], "reader and writer assigned to readers");
+});
+
+test("DuplexCollection: _byGlob", (t) => {
+	t.plan(2);
+
+	const abstractReader = {
+		_byGlob: sinon.stub().returns(Promise.resolve([]))
+	};
+
+	const duplexCollection = new DuplexCollection({
+		name: "myCollection",
+		reader: abstractReader,
+		writer: abstractReader
+	});
+
+	const trace = {
+		collection: sinon.spy()
+	};
+
+	const comboSpy = sinon.spy(duplexCollection._combo, "_byGlob");
+
+	return duplexCollection._byGlob("anyPattern", {someOption: true}, trace)
+		.then(function(resources) {
+			t.true(Array.isArray(resources), "Found resources are returned as an array");
+			t.true(comboSpy.calledWithExactly("anyPattern", {someOption: true}, trace),
+				"Delegated globbing task correctly to readers");
+		});
+});
+
+test("DuplexCollection: _byPath with reader finding a resource", (t) => {
+	t.plan(4);
+
+	const resource = new Resource({
+		path: "path",
+		buffer: new Buffer("content")
+	});
+
+	const pushCollectionSpy = sinon.spy(resource, "pushCollection");
+
+	const abstractReader = {
+		_byPath: sinon.stub().returns(Promise.resolve(resource))
+	};
+
+	const trace = {
+		collection: sinon.spy()
+	};
+
+	const duplexCollection = new DuplexCollection({
+		name: "myCollection",
+		reader: abstractReader,
+		writer: abstractReader
+	});
+
+	const comboSpy = sinon.spy(duplexCollection._combo, "_byPath");
+
+	return duplexCollection._byPath("anyVirtualPath", {someOption: true}, trace)
+		.then(function(resource) {
+			t.true(comboSpy.calledWithExactly("anyVirtualPath", {someOption: true}, trace),
+				"Delegated globbing task correctly to readers");
+			t.true(pushCollectionSpy.called, "pushCollection called on resource");
+			resource.getString().then(function(content) {
```

Comment:
There's a missing `return` here, right? Otherwise the deepEqual assert below might be executed after the test has finished, right?

Reply:
Done.